### PR TITLE
Do not use ZeroTensor for inplace ops

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70355
* #69956
* __->__ #69998
* #69884

Fixes: https://github.com/pytorch/pytorch/issues/69855

The check for undefined grads for forward AD was not being run because `check_undefined_grads` was only passed as True by OpInfo for backward AD. This PR updates gradcheck to interpret `check_undefined_grads` as possibly for forward or backward AD.

This PR also updates codegen to 1) not use ZeroTensor for `self` when the op is inplace. 2) only create zeros (either through ZeroTensor or at::zeros) if the tensor itself is not undefined. Previously we would error in this case when we call `.options` on the undefined tensor.

~TODO: undo the skips that are due to the original issue~

Differential Revision: [D33235973](https://our.internmc.facebook.com/intern/diff/D33235973)

### CODE REVIEW

**Comment by anjali411** [comment]
I don't think this is what we want to do. We should fix the derivative formulas where inplace ops are beign called and ensure that we materialize the ZT before we call them!

**Comment by albanD** [comment]
@anjali411 that is a good question.
My idea here was that, because all inplace op will have this issue and the fix for all of them will be the same, it is nice for the codegen to do it for you.

**Comment by anjali411** [comment]
> @anjali411 that is a good question. My idea here was that, because all inplace op will have this issue and the fix for all of them will be the same, it is nice for the codegen to do it for you.

That sounds good. I'd suggest only materializing the first ZT gradient (for self), since we are guaranteed to only have the first argument as a mutable argument for an inplace op.

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -908,12 +909,13 @@ def emit_fw_derivatives() -> List[str]:
 
             unpacked_arguments = ""
             for inp in differentiable_inputs:
+                zeros_fn = "zeros" if inplace and inp.name == "self" else "_efficientzerotensor"
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:912]: nit: you should not hardcode "self" here but use the mapping defined in: https://github.com/pytorch/pytorch/blob/92463573d87195f87f68f32dbb48ee497cdea182/tools/autograd/gen_inplace_or_view_type.py#L42

Comment (soulitzer) [reply tools/autograd/gen_variable_type.py:912]: Hmm, I don't completely understand this case yet. Doesn't this mapping tells us which input the output is aliased with? For inplace ops my understanding that they are always methods and the mutable input will automatically be `self` no?

Comment (albanD) [reply tools/autograd/gen_variable_type.py:912]: Ho true this is for views.
Not sure if inplace is always on self...

Comment (anjali411) [reply tools/autograd/gen_variable_type.py:912]: yup @soulitzer is right, inplace ops always have `self` as the mutable input

**Review by albanD — APPROVED** [review]
Ok then

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33235973).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33235973).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33235973).

**Pull request closed without merging**